eZ/Bundle/EzPublishCoreBundle/Fragment/DirectFragmentRenderer.php

```diff
@@ -86,4 +99,69 @@ protected function getArguments(callable $controller, ControllerEvent $event): a
 
         return $arguments;
     }
+
+    protected function getFragmentUri(
+        ControllerReference $reference,
+        Request $request
+    ): string {
+        $simplifiedReference = clone $reference;
+        $simplifiedReference->attributes = [
+            '_format' => $reference->attributes['_format'] ?? null,
+            '_locale' => $reference->attributes['_locale'] ?? null,
+        ];
+
+        return $this->generateFragmentUri($simplifiedReference, $request, false, false);
+    }
+
+    private function getSubRequest(Request $request): Request
+    {
+    }
+
+    /**
+     * @param string|\Symfony\Component\HttpKernel\Controller\ControllerReference $uri
+     *
+     * @throws \eZ\Publish\Core\MVC\Symfony\Templating\Exception\InvalidResponseException
+     */
+    public function render(
+        $uri,
+        Request $request,
+        array $options = []
+    ): Response {
+        $controllerReference = null;
+        if ($uri instanceof ControllerReference) {
+            $controllerReference = $uri;
+            $uri = $this->getFragmentUri($controllerReference, $request);
+        }
+
+        $subRequest = $this->createSubRequest($uri, $request);
+        $subRequest->attributes->set('siteaccess', $request->attributes->get('siteaccess'));
+
+        if (null !== $controllerReference) {
+            $subRequest->attributes->add($controllerReference->attributes);
+            $subRequest->attributes->set('_controller', $controllerReference->controller);
+        }
+
+        $event = $this->getControllerEvent($subRequest);
+        $controller = $this->getController($event);
+        $arguments = $this->getArguments($controller, $event);
+
+        $response = $controller(...array_values($arguments));
+
+        if ($response instanceof Response) {
+            return $response;
+        } elseif ($response instanceof View) {
+            return new Response($this->viewTemplateRenderer->render($response));
+        } elseif (is_string($response)) {
+            return new Response($response);
+        }
+
+        throw new InvalidResponseException(
+            sprintf('Unsupported type (%s)', Php80::get_debug_type($response))
```

Comment:
Just call it exactly as you would in PHP8, otherwise there's no point in using a polyfill.

```suggestion
            sprintf('Unsupported type (%s)', \get_debug_type($response))
```

Reply:
And, if we're not having `symfony/polyfill-php80` in `composer.json`, then we cannot use it. I was just noting that it's a possibility to use those polyfills.